nyaa/api_torznab.py

```diff
@@ -101,71 +112,113 @@ def torznab_caps():
-    rid = flask.request.args.get('rid')
+    query_term = api_request.args.get('q')
+    query_season = api_request.args.get('season')
+    query_episode = api_request.args.get('ep')
+    query_rageid = api_request.args.get('rid')
 
-    cat = flask.request.args.get('cat', default='100000')
+    if query_season is not None or query_episode is not None or query_rageid is not None:
+        return render_torznab_error(201)
 
-    # Quality filter (remake/trusted) has no predefined query param yet. Propose 'tag' as generic cross newznab/torznab field to remain site agnostic.
-    #quality_filter = flask.request.args.get('f')
-    extended = flask.request.args.get('extended')
-    offset = flask.request.args.get('offset', default=0, type=int)
-    limit = flask.request.args.get('limit', default=page_size, type=int)
+    cat = api_request.args.get('cat', default='')
+    # Quality filter (remake/trusted) has no predefined query param yet.
+    # Propose 'tag' as generic cross newznab/torznab field to remain site agnostic.
+    tag = api_request.args.get('tag', default='')
+    extended = api_request.args.get('extended', False, type=bool)
+    offset = api_request.args.get('offset', default=0, type=int)
+    limit = api_request.args.get('limit', default=page_size, type=int)
 
-    if season or ep or rid:
-        return render_torznab_error(201)
+    try:
+        limit = min(limit, page_size)
+        if limit == 0:
+            raise EmptySearchError('Limit of 0 guaranteed guarantees empty result')
+        if offset % limit:
+            app.logger.warn('offset {} isn\'t a multiple of the page size {}'.format(offset, limit))
+            return render_torznab_error(201)
+
+        categories = get_categories_filter(cat)
+
+        (include_tags, exclude_tags) = get_tag_filter(tag)
 
-    page = int(1 + offset / page_size)
-    # TODO: Cleanup q
-    # TODO: category can have multiple, yet search does not support that.
-    category = None
-    quality_filter = None
-    # TODO: offset = usually per page, but not guaranteed, give error if not a multiple of page size.
-    #       Or fix search module to accept offset & limit instead of page.
-    # TODO: maximum of limit should be the app.config['RESULTS_PER_PAGE']
-    query_args = {
-        'term': q or '',
-        'category': category or '0_0',
-        'quality_filter': quality_filter or '0',
-        'page': page or 1,
-        'rss': False  # We want it to page
-    }
+        # TODO: Cleanup q
 
-    # God mode
-    if flask.g.user and flask.g.user.is_admin:
-        query_args['admin'] = True
+        query_args = {
+            'term': query_term,
+            'categories': categories,
+            'include_tags': include_tags,
+            'exclude_tags': exclude_tags,
+            'page': 1 + int(offset / limit),
+            'page_size': limit
+        }
 
-    query = search(**query_args)
+        query_results = search_custom(**query_args)
+    except EmptySearchError:
```

Comment:
The EmptySearchError is raised in various locations inside the search operation, including inside methods that are not part of the flow that returns the search result, and thus would not be able to do what you suggest.
Ironically I thought _this_ would be the more pythonic way of doing things, throwing the error and leave a higher up to decide whether to raise 404 or an empty result. If that's wrong, then fine, but please tell me how to do methods like `get_categories_filter`  and `get_tag_filter`.